starlark/testdata/bytes.star

```diff
@@ -0,0 +1,112 @@
+hello = bytes("hello, 世界")
+goodbye = bytes("goodbye")
+empty = bytes("")
+nonprinting = bytes("\t\n\u200D")  # TAB, NEWLINE, DEL, ZERO_WIDTH_JOINER
+assert.eq(bytes("hello, 世界"[:-1]), b"hello, 世��")
+
+# literals
+assert.eq(b"hello, 世界", hello)
+assert.eq(b"goodbye", goodbye)
+assert.eq(b"", empty)
+assert.eq(b"\t\n\u200D", nonprinting)
+assert.ne("abc", b"abc")
+assert.eq(b"\012\xff\u0400\U0001F63F", b"\n\xffЀ😿") # see scanner tests for more
+assert.eq(rb"\r\n\t", b"\\r\\n\\t") # raw
+
+# type
+assert.eq(type(hello), "bytes")
+
+# len
+assert.eq(len(hello), 13)
+assert.eq(len(goodbye), 7)
+assert.eq(len(empty), 0)
+
+# truth
+assert.true(hello)
+assert.true(goodbye)
+assert.true(not empty)
+
+# str(bytes) does UTF-8 to UTF-k transcoding.
+assert.eq(str(hello), "hello, 世界")
+assert.eq(str(hello[:-1]), "hello, 世��")  # incomplete UTF-8 encoding => U+FFFD
+assert.eq(str(goodbye), "goodbye")
+assert.eq(str(empty), "")
+assert.eq(str(nonprinting), "\t\n\u200d")
+assert.eq(str(b"\udc00"), "�") # unpaired surrogate => U+FFFD
+
+# reprq
+assert.eq(repr(hello), r'b"hello, 世界"')
+assert.eq(repr(hello[:-1]), r'b"hello, 世\xe7\x95"')  # (incomplete UTF-8 encoding )
+assert.eq(repr(goodbye), 'b"goodbye"')
+assert.eq(repr(empty), 'b""')
+assert.eq(repr(nonprinting), 'b"\\t\\n\\u200d"')
+
+# equality
+assert.eq(hello, hello)
+assert.ne(hello, goodbye)
+assert.eq(bytes("goodbye"), goodbye)
+
+# ordered comparison
+assert.lt(bytes("abc"), bytes("abd"))
+assert.lt(bytes("abc"), bytes("abcd"))
+assert.lt(bytes("\x7f"), bytes("\x80")) # bytes compare as uint8, not int8
+
+# bytes are dict-hashable
+dict = {hello: 1, goodbye: 2}
+dict[bytes("goodbye")] = 3
+assert.eq(len(dict), 2)
+assert.eq(dict[goodbye], 3)
+
+# indexing
+assert.eq(goodbye[0], bytes("g"))
+assert.eq(goodbye[-1], bytes("e"))
+assert.fails(lambda: goodbye[100], "out of range")
+
+# slicing
+assert.eq(goodbye[:4], bytes("good"))
+assert.eq(goodbye[4:], bytes("bye"))
+assert.eq(goodbye[::2], bytes("gobe"))
+assert.eq(goodbye[3:4], bytes("d"))  # special case: len=1
+assert.eq(goodbye[4:4], bytes(""))  # special case: len=0
+
+# ord
+assert.eq(ord(b"a"), 97)
+assert.fails(lambda: ord(b"ab"), "ord: bytes has length 2, want 1")
+assert.fails(lambda: ord(b""), "ord: bytes has length 0, want 1")
+
+# repeat (bytes * int)
+assert.eq(goodbye * 3, bytes("goodbyegoodbyegoodbye"))
+assert.eq(3 * goodbye, bytes("goodbyegoodbyegoodbye"))
+
+# elems() returns an iterable value over 1-byte substrings.
+assert.eq(type(hello.elems()), "bytes.elems")
+assert.eq(str(hello.elems()), "b\"hello, 世界\".elems()")
+assert.eq(list(hello.elems()),
+          [b"h", b"e", b"l", b"l", b"o", b",", b" ",
+           b"\xE4", b"\xB8", b"\x96", b"\xE7", b"\x95", b"\x8C"])
+assert.eq(list(goodbye.elems()),
+          [b"g", b"o", b"o", b"d", b"b", b"y", b"e"])
+assert.eq(list(empty.elems()), [])
+
+# TODO(adonovan);
+# - more methods: find, index, split, etc.
+# - constructor from int values: bytes([1, 2, 3])?
```

Comment:
Jay, do you have any thoughts on what should be the syntax to construct a bytes from a list of numeric byte values, or the various ord/chr operations to convert numbers to/from bytes, UTF-k codes, and Unicode code points?

Reply:
I think Starlark should match Python 3 wherever possible. I don't know Python 3 that well though.

For the `bytes` constructor, I think that means accepting a string, `bytes`, or sequence of integers in the range [0, 256). We probably shouldn't support the `encoding` or `errors` parameters for now.

`ord` seems correct: we should only accept `bytes` of length 1.

`chr` I think should convert a code point to however many UTF-k codes are needed? `bytes(chr(i))` would convert to UTF-8 I think? I'm probably missing some cases though.

Would `str(bytes(chr(i))` be equivalent to `chr(i)`?

Reply:
Ah, I didn't know that bytes([int]) worked; good. I've implemented that. I also made the string.elems iterable view indexable, which addresses some of the gaps.

The only missing case now is how to create a 1-element string, or a string from a list of int values, analogous to bytes([int]). Any ideas?

Take a look at this updated comment for a summary of the various operations.

> Would str(bytes(chr(i)) be equivalent to chr(i)?

Correct.

Reply:
> The only missing case now is how to create a 1-element string, or a string from a list of int values, analogous to bytes([int]). Any ideas?

I suppose bytes([int]) is good enough for now, and we can optimize it later.